File: android/app/src/main/kotlin/ai/gowtam/oak/features/artifact/PokemonCompare.kt
package ai.gowtam.oak.features.artifact

import ai.gowtam.oak.wire.Abilities
import ai.gowtam.oak.wire.BaseStats
import ai.gowtam.oak.wire.Format
import ai.gowtam.oak.wire.OffensiveProfile
import ai.gowtam.oak.wire.PokemonArtifactData
import ai.gowtam.oak.wire.TeamMember

/**
 * Client-side two-subject Pokémon compare (ADR-11).
 *
 * No compare endpoint — the caller fetches both `/api/entity` profiles
 * and this helper diffs them. Cross-scope pairs keep both format tags
 * (CMP-BR-2). Not a chat turn (CMP-BR-4).
 */

/** Stated default when neither subject carries a set (CMP-AC-3.2). */
private const val DEFAULT_SPEED_LEVEL = 50

data class PokemonCompareSubject(
    val format: Format,
    val profile: PokemonArtifactData,
    val set: TeamMember? = null,
    val offensive: OffensiveProfile? = null,
)

data class PokemonCompareSetDiff(
    val onlyLeft: List<String>,
    val onlyRight: List<String>,
    val shared: List<String>,
)

data class PokemonCompareSide(
    val format: Format,
    val displayName: String,
)

data class PokemonCompareStatsPair(
    val left: BaseStats,
    val right: BaseStats,
)

data class PokemonCompareSpeed(
    val leftValue: Int,
    val rightValue: Int,
    val defaultLevel: Int,
    val usedSetLeft: Boolean,
    val usedSetRight: Boolean,
)

data class PokemonCompareMatchups(
    val weakTo: PokemonCompareSetDiff,
    val resists: PokemonCompareSetDiff,
    val immuneTo: PokemonCompareSetDiff,
    val offensiveSuperEffective: PokemonCompareSetDiff,
)

data class PokemonCompareDiff(
    val left: PokemonCompareSide,
    val right: PokemonCompareSide,
    val stats: PokemonCompareStatsPair,
    val types: PokemonCompareSetDiff,
    val abilities: PokemonCompareSetDiff,
    val speed: PokemonCompareSpeed,
    val movepool: PokemonCompareSetDiff,
    val matchups: PokemonCompareMatchups,
)

/** Diff two portable profiles. Does not mutate inputs. */
fun diffPokemonProfiles(
    left: PokemonCompareSubject,
    right: PokemonCompareSubject,
): PokemonCompareDiff {
    val leftAbilities = abilitySlugs(left.profile.abilities)
    val rightAbilities = abilitySlugs(right.profile.abilities)
    val usedSetLeft = left.set != null
    val usedSetRight = right.set != null

    return PokemonCompareDiff(
        left = PokemonCompareSide(format = left.format, displayName = left.profile.displayName),
        right = PokemonCompareSide(format = right.format, displayName = right.profile.displayName),
        stats = PokemonCompareStatsPair(
            left = left.profile.baseStats,
            right = right.profile.baseStats,
        ),
        types = setDiff(left.profile.types, right.profile.types),
        abilities = setDiff(leftAbilities, rightAbilities),
        speed = PokemonCompareSpeed(
            leftValue = left.profile.baseStats.spe,
            rightValue = right.profile.baseStats.spe,
            defaultLevel = DEFAULT_SPEED_LEVEL,
            usedSetLeft = usedSetLeft,
            usedSetRight = usedSetRight,
        ),
        movepool = setDiff(movepoolSlugs(left.profile), movepoolSlugs(right.profile)),
        matchups = PokemonCompareMatchups(
            weakTo = setDiff(left.profile.matchups.weakTo, right.profile.matchups.weakTo),
            resists = setDiff(left.profile.matchups.resists, right.profile.matchups.resists),
            immuneTo = setDiff(left.profile.matchups.immuneTo, right.profile.matchups.immuneTo),
            offensiveSuperEffective = setDiff(
                left.offensive?.superEffectiveAgainst.orEmpty(),
                right.offensive?.superEffectiveAgainst.orEmpty(),
            ),
        ),
    )
}

private fun abilitySlugs(abilities: Abilities): List<String> {
    val slugs = mutableListOf<String>()
    if (abilities.slot1.isNotEmpty()) slugs += abilities.slot1
    abilities.slot2?.takeIf { it.isNotEmpty() }?.let { slugs += it }
    abilities.hidden?.takeIf { it.isNotEmpty() }?.let { slugs += it }
    return slugs
}

private fun movepoolSlugs(profile: PokemonArtifactData): List<String> =
    profile.movepool.flatMap { group -> group.moves.map { it.slug } }

private fun setDiff(left: List<String>, right: List<String>): PokemonCompareSetDiff {
    val leftSet = left.toSet()
    val rightSet = right.toSet()
    val onlyLeft = mutableListOf<String>()
    val onlyRight = mutableListOf<String>()
    val shared = mutableListOf<String>()
    val seen = mutableSetOf<String>()

    for (item in left) {
        if (!seen.add(item)) continue
        if (item in rightSet) shared += item else onlyLeft += item
    }
    for (item in right) {
        if (item in seen || item in leftSet) continue
        seen += item
        onlyRight += item
    }
    return PokemonCompareSetDiff(onlyLeft = onlyLeft, onlyRight = onlyRight, shared = shared)
}

File: ios/OakApp/Features/Artifact/PokemonCompare.swift
import Foundation

/// Client-side two-subject Pokémon compare (ADR-11).
///
/// Clones `web/src/lib/pokemon-compare.ts` against the native artifact + set
/// types. No compare endpoint — the caller fetches both `/api/entity` profiles
/// and this helper diffs them. Cross-scope pairs keep both format tags
/// (CMP-BR-2). Not a chat turn (CMP-BR-4).
///
/// Movepool slugs are flattened across learn-method groups. Ability slugs are
/// the non-nil `slot1` / `slot2` / `hidden` values. Offensive matchup-diff
/// reads `subject.offensive` (no type-chart port). Speed states a default
/// level unless `set` is present (CMP-AC-3.2).

/// Stated default when neither subject carries a set (CMP-AC-3.2).
private let defaultSpeedLevel = 50

struct PokemonCompareSubject: Sendable {
  let format: Format
  let profile: PokemonArtifactData
  let set: TeamMember?
  let offensive: OffensiveProfile?
}

struct PokemonCompareSetDiff: Equatable, Sendable {
  let onlyLeft: [String]
  let onlyRight: [String]
  let shared: [String]
}

struct PokemonCompareSide: Equatable, Sendable {
  let format: Format
  let displayName: String
}

struct PokemonCompareStatsPair: Equatable, Sendable {
  let left: BaseStats
  let right: BaseStats
}

struct PokemonCompareSpeed: Equatable, Sendable {
  let leftValue: Int
  let rightValue: Int
  let defaultLevel: Int
  let usedSetLeft: Bool
  let usedSetRight: Bool
}

struct PokemonCompareMatchups: Equatable, Sendable {
  let weakTo: PokemonCompareSetDiff
  let resists: PokemonCompareSetDiff
  let immuneTo: PokemonCompareSetDiff
  let offensiveSuperEffective: PokemonCompareSetDiff
}

struct PokemonCompareDiff: Equatable, Sendable {
  let left: PokemonCompareSide
  let right: PokemonCompareSide
  let stats: PokemonCompareStatsPair
  let types: PokemonCompareSetDiff
  let abilities: PokemonCompareSetDiff
  let speed: PokemonCompareSpeed
  let movepool: PokemonCompareSetDiff
  let matchups: PokemonCompareMatchups
}

/// Diff two portable profiles. Does not mutate inputs.
func diffPokemonProfiles(
  _ left: PokemonCompareSubject,
  _ right: PokemonCompareSubject
) -> PokemonCompareDiff {
  let leftAbilities = abilitySlugs(left.profile.abilities)
  let rightAbilities = abilitySlugs(right.profile.abilities)
  let leftOffensive = left.offensive?.superEffectiveAgainst ?? []
  let rightOffensive = right.offensive?.superEffectiveAgainst ?? []

  return PokemonCompareDiff(
    left: PokemonCompareSide(format: left.format, displayName: left.profile.displayName),
    right: PokemonCompareSide(format: right.format, displayName: right.profile.displayName),
    stats: PokemonCompareStatsPair(
      left: left.profile.baseStats,
      right: right.profile.baseStats
    ),
    types: setDiff(left.profile.types, right.profile.types),
    abilities: setDiff(leftAbilities, rightAbilities),
    speed: PokemonCompareSpeed(
      leftValue: left.profile.baseStats.spe,
      rightValue: right.profile.baseStats.spe,
      defaultLevel: defaultSpeedLevel,
      usedSetLeft: left.set != nil,
      usedSetRight: right.set != nil
    ),
    movepool: setDiff(flattenMovepool(left.profile.movepool), flattenMovepool(right.profile.movepool)),
    matchups: PokemonCompareMatchups(
      weakTo: setDiff(left.profile.matchups.weakTo, right.profile.matchups.weakTo),
      resists: setDiff(left.profile.matchups.resists, right.profile.matchups.resists),
      immuneTo: setDiff(left.profile.matchups.immuneTo, right.profile.matchups.immuneTo),
      offensiveSuperEffective: setDiff(leftOffensive, rightOffensive)
    )
  )
}

private func abilitySlugs(_ abilities: Abilities) -> [String] {
  var slugs: [String] = []
  if !abilities.slot1.isEmpty { slugs.append(abilities.slot1) }
  if let slot2 = abilities.slot2, !slot2.isEmpty { slugs.append(slot2) }
  if let hidden = abilities.hidden, !hidden.isEmpty { slugs.append(hidden) }
  return slugs
}

private func flattenMovepool(_ groups: [MovepoolGroup]) -> [String] {
  groups.flatMap { $0.moves.map(\.slug) }
}

private func setDiff(_ left: [String], _ right: [String]) -> PokemonCompareSetDiff {
  let leftSet = Set(left)
  let rightSet = Set(right)
  var onlyLeft: [String] = []
  var onlyRight: [String] = []
  var shared: [String] = []
  var seen = Set<String>()

  for item in left {
    if seen.contains(item) { continue }
    seen.insert(item)
    if rightSet.contains(item) {
      shared.append(item)
    } else {
      onlyLeft.append(item)
    }
  }
  for item in right {
    if seen.contains(item) || leftSet.contains(item) { continue }
    seen.insert(item)
    onlyRight.append(item)
  }
  return PokemonCompareSetDiff(onlyLeft: onlyLeft, onlyRight: onlyRight, shared: shared)
}
